go/obscuronode/enclave/sql/edgelessdb.go

```diff
@@ -0,0 +1,480 @@
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"database/sql"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/edgelesssys/ego/ecrypto"
+
+	"github.com/obscuronet/obscuro-playground/go/log"
+
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/go-sql-driver/mysql"
+)
+
+/*
+   Getting an edgeless DB connection is a bit of a dance in order to achieve mutual trust.
+
+   The edgeless DB is unusable when it's first started, as an owner, we must initially:
+     - do a remote attestation on the report provided by {edbAddress}/quote
+     - create a ca cert that will authenticate our DB users going forwards
+     - prepare a manifest.json that contains that CA cert and some SQL to initialise the DB tables and user
+     - submit that manifest.json file to {edbAddress}/manifest, using the certificate provided from /quote to authenticate
+     - seal and persist the manifest.json and the certs so we can reconnect if enclave is restarted
+
+   When connecting to an already-initialized edgeless DB we must:
+     - perform remote attestation on the edgeless db
+     - unseal the manifest.json and get the hash of it, also unseal the certificate that edb was initialized with
+	 - verify the /signature request for edgeless DB matches the manifest.json hash
+     - connect to edb with the persisted cert - it's now safe to read and write to the DB
+
+	Some useful documentation for this:
+		Main edb docs: https://docs.edgeless.systems/edgelessdb/#/
+		EDB demo docs: https://github.com/edgelesssys/edgelessdb/tree/main/demo
+		// Note: due to an issue with the dependency, I've duplicated the relevant parts of the ERA tool into edbattestation.go
+		ERA - remote attestation tool: https://github.com/edgelesssys/era
+*/
+
+var manifestSQLStatements = []string{
+	"CREATE USER obscuro REQUIRE ISSUER '/CN=obscuroCA' SUBJECT '/CN=obscuroUser'",
+	"CREATE DATABASE obsdb",
+	"CREATE TABLE obsdb.keyvalue (ky varbinary(64) primary key, val blob)",
+	"GRANT ALL ON obsdb.keyvalue TO obscuro",
+}
+
+const (
+	dataDir                = "/data/"
+	manifestFilepath       = dataDir + "manifest.json"
+	caCertFilepath         = dataDir + "ca-cert.pem"
+	userCertFilepath       = dataDir + "user-cert.pem"
+	userKeyFilepath        = dataDir + "user-key.pem"
+	attestationCfgFilepath = dataDir + "edgelessdb-sgx.json"
+	edbHTTPPort            = "8080"
+)
+
+type manifest struct {
+	SQL   []string `json:"sql"`
+	Cert  string   `json:"ca"`
+	Debug bool     `json:"debug"`
+}
+
+type EdgelessDBConfig struct {
+	Host string
+}
+
+func EdgelessDBConnector(edbCfg EdgelessDBConfig) (ethdb.Database, error) {
+	// before we try to connect to the Edgeless DB we have to do remote attestation on it
+	edbPEM, err := performRemoteAttestation(edbCfg.Host)
+	if err != nil {
+		return nil, fmt.Errorf("remote attestation of edgeless DB failed - %w", err)
+	}
+	log.Info("retrieved edb PEM: %s", edbPEM)
```

Comment:
@BedrockSquirrel Is it safe to log the PEM?

And in general, I think we should avoid the acronym `edb` in our logs. It's lazy and will be confusing to our users.